runtime/drive_handler.go

```diff
@@ -0,0 +1,182 @@
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	// fcSectorSize is the sector size of Firecracker
+	fcSectorSize = 512
+)
+
+var (
+	// ErrDrivesExhausted occurs when there are no more drives left to use. This
+	// can happen by calling PatchStubDrive greater than the number of drives.
+	ErrDrivesExhausted = fmt.Errorf("There are no remaining drives to be used")
+
+	// ErrDriveIDNil should never happen, but we safe guard against nil dereferencing
+	ErrDriveIDNil = fmt.Errorf("DriveID of current drive is nil")
+)
+
+// stubDriveHandler is used to manage stub drives.
+type stubDriveHandler struct {
+	RootPath       string
+	stubDriveIndex int64
+	drives         []models.Drive
+	logger         *logrus.Entry
+	mutex          sync.Mutex
+}
+
+func newStubDriveHandler(path string, logger *logrus.Entry) stubDriveHandler {
+	return stubDriveHandler{
+		RootPath: path,
+	}
+}
+
+// StubDrivePaths will create stub drives and return the paths associated with
+// the stub drives.
+func (h *stubDriveHandler) StubDrivePaths(count int) ([]string, error) {
+	paths := []string{}
+	for i := 0; i < count; i++ {
+		driveID := fmt.Sprintf("stub%d", i)
+		path := filepath.Join(h.RootPath, driveID)
+
+		if err := h.createStubDrive(driveID, path); err != nil {
+			return nil, err
+		}
+
+		paths = append(paths, path)
+	}
+
+	return paths, nil
+}
+
+func (h *stubDriveHandler) createStubDrive(driveID, path string) error {
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			h.logger.WithError(err).Errorf("unexpected error during %v close", f.Name())
+		}
+	}()
+
+	stubContent, err := internal.GenerateStubContent(driveID)
+	if err != nil {
+		return err
+	}
+
+	if _, err := f.WriteString(stubContent); err != nil {
+		return err
+	}
+
+	info, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	fileSize := info.Size()
+	sectorAmount := fileSize / fcSectorSize
+	truncateSize := fcSectorSize * sectorAmount
+
+	remainingBytes := fileSize % fcSectorSize
+	if remainingBytes != 0 {
+		// If there are any residual bytes, this means we've need to fill the
+		// appropriate sector size to ensure that the data is visible to
+		// Firecracker.
+		truncateSize += fcSectorSize
+	}
+
+	// Firecracker will not show any drives smaller than 512 bytes. In
+	// addition, the drive is read in chunks of 512 bytes; if the drive size is
+	// not a multiple of 512 bytes, then the remainder will not be visible to
+	// Firecracker. So we truncate to the appropriate size based on the residual
```

Comment:
This language is somewhat confusing, since "truncate" means to "chop off", but that's not precisely what we're doing. I think it'd be clearer if we used `driveSize` or something to that effect instead of `truncateSize` (see [dict.org](http://www.dict.org/bin/Dict?Form=Dict2&Database=*&Query=truncate) for definitions, and yes, POSIX and consequently Go use the word in a confusing way.)